File: April 2022/April_30.java
class Solution {
    class Pair{
        String end;
        double val;
        Pair(String a, double b){
            end = a;
            val = b;
        }
    }
    
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        Map<String, List<Pair>> map = new HashMap<>();
        for(int i = 0; i < equations.size(); i++){
            if(!map.containsKey(equations.get(i).get(0)))
                map.put(equations.get(i).get(0), new ArrayList<>());
            if(!map.containsKey(equations.get(i).get(1)))
                map.put(equations.get(i).get(1), new ArrayList<>());
            map.get(equations.get(i).get(0)).add(new Pair(equations.get(i).get(1), values[i]));
            map.get(equations.get(i).get(1)).add(new Pair(equations.get(i).get(0), 1 / values[i]));
        }
        
        int n = queries.size();
        double[] ans = new double[n];
        for(int i = 0; i < n; i++){
            if(!map.containsKey(queries.get(i).get(0)) || !map.containsKey(queries.get(i).get(1))){
                ans[i] = -1.0;
                continue;
            }
            List<Pair> l1 = new ArrayList<>(), l2 = new ArrayList<>();
            Set<String> vis = new HashSet<>();
            find(queries.get(i).get(0), 1.0 , map, l1, vis);
            find(queries.get(i).get(1), 1.0 , map, l2, vis);            
            boolean flag = false;
            for(Pair p1 : l1){
                for(Pair p2 : l2)
                    if(p1.end.equals(p2.end)){
                        ans[i] = p1.val / p2.val;
                        flag = true;
                        break;
                    }
                if(flag)
                    break;
            }
            if(!flag)
                ans[i] = -1.0;
        }
        return ans;
    }
    
    private void find(String s, double d, Map<String, List<Pair>> map, List<Pair> list, Set<String> vis){
        vis.add(s);
        list.add(new Pair(s, d));
        for(Pair nbr : map.get(s))
            if(!vis.contains(nbr.end))
                find(nbr.end, d * nbr.val, map, list, vis);
        vis.remove(s);
    }
}

File: April 2022/April_30.cpp
class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, vector<pair<string, double> > > mp;
        for(int i = 0; i < equations.size(); i++){
            mp[equations[i][0]].push_back({equations[i][1], values[i]});
            mp[equations[i][1]].push_back({equations[i][0], 1 / values[i]});
        }
        
        int n = queries.size();
        vector<double> ans(n, -1.0);
        for(int i = 0; i < n; i++){
            if(mp.find(queries[i][0]) != mp.end() && mp.find(queries[i][1]) != mp.end()){
                vector<pair<string, double>> l1, l2;
                unordered_set<string> vis;
                find(queries[i][0], 1.0, mp, l1, vis);
                find(queries[i][1], 1.0, mp, l2, vis);
                bool flag = false;
                for(auto &p1 : l1){
                    for(auto &p2 : l2){
                        if(p1.first == p2.first){
                            ans[i] = p1.second / p2.second;
                            flag = true;
                            break;
                        }
                    }
                    if(flag)
                        break;
                }
            }
        }
        return ans;
    }
    
    void find(string &s, double d, unordered_map<string, vector<pair<string, double> > > &mp, vector<pair<string, double> > & vec, unordered_set<string> vis){
        vis.insert(s);
        for(auto &p : mp[s])
            if(vis.find(p.first) == vis.end())
                find(p.first, d * p.second, mp, vec, vis);
        vis.erase(s);
        vec.push_back({s, d});
    }
};
